main.c

```diff
@@ -0,0 +1,572 @@
+    if (len_size == -1)
+        return -1;
+    if (len_size < max_size) {
+        input[len_size - 1] = '\0';
+        strncpy(dst, input, len_size);
+    }
+    else {
+        input[max_size - 1] = '\0';
+        strncpy(dst, input, max_size);
+        len_size = max_size;
+    }
+    free(input);
+    return len_size;
+}
+
+// Function to read students from file and build the school array
+
+void readStudents(FILE* filePointer) {
+    Student student;
+    int level, class;
+    while (fscanf(filePointer, "%s %s %s %d %d", student.firstName, student.lastName, student.phoneNumber, &level, &class) == 5) {
+        // Check if level and class are within valid bounds
+        if (level >= 1 && level <= LEVELS && class >= 1 && class <= CLASS) {
+            int phoneNumberLength = strlen(student.phoneNumber);
+            int containsOnlyDigits = 1;
+             if(phoneNumberLength>MAXNUMP)
+                containsOnlyDigits=0;
+
+            for (int i = 0; i < phoneNumberLength; i++) {
+                if (!isdigit(student.phoneNumber[i])) {
+                    containsOnlyDigits = 0;
+                    break;
+                }
+            }
+
+            if (!containsOnlyDigits) {
+                printf("Invalid phone number format for student: %s %s\n", student.firstName, student.lastName);
+                continue;
+            }
+
+            int firstNameLength = strlen(student.firstName);
+            if(strlen(student.firstName)>MAXNUMF){
+                printf("firstname too long for student with phone number: %s\n", student.phoneNumber);
+                continue;
+            }
+
+            int lastNameLength = strlen(student.lastName);
+            if(strlen(student.lastName)>MAXNUML){
+                printf("lastName too long for student with phone number: %s\n", student.phoneNumber);
+                continue;
+            }
+            // Dynamically allocate memory for the student
+            Student* newStudent = (Student*)malloc(sizeof(Student));
+            if (newStudent == NULL) {
+                printf("Memory allocation failed.\n");
+                exit(1);
+            }
+
+            // Copy the read student data to the newly allocated memory
+            strcpy(newStudent->firstName, student.firstName);
+            strcpy(newStudent->lastName, student.lastName);
+            strcpy(newStudent->phoneNumber, student.phoneNumber);
+
+            // Read the grades from the file for the current student
+            for (int i = 0; i < 10; i++) 
+                fscanf(filePointer, "%d", &newStudent->grades[i]);
+            
+            newStudent->next = NULL;
+
+            // Add the newStudent to the end of the linked list for the specific level and class
+            if (school[level - 1][class - 1] == NULL) 
+                school[level - 1][class - 1] = newStudent;
+             else {
+                Student* p = school[level - 1][class - 1];
+                while (p->next != NULL) 
```

Comment:
Adding a new student to a 10 million linked list of students, will you go all the way to the end of the list
to add the new student? Think of an efficient way to insert a new student into the list.